File: sample-conference-videochat/sample-conference-videochat/Components/Auth/Views/LoadingButton/QBLoadingButton.m
//
//  QBLoadingButton.m
//  LoginComponent
//
//  Created by Andrey Ivanov on 01/06/16.
//  Copyright © 2016 Quickblox. All rights reserved.
//

#import "QBLoadingButton.h"

@interface QBLoadingButton()

@property (strong, nonatomic, readonly) CAShapeLayer *shapeLayer;
@property (strong, nonatomic) UIActivityIndicatorView *activity;
@property (strong, nonatomic) NSString *currentText;

@end

@implementation QBLoadingButton

- (void)awakeFromNib {
    
    [super awakeFromNib];
    
    self.shapeLayer.fillColor = [[UIColor colorWithRed:0.0392 green:0.3765 blue:1.0 alpha:1.0] CGColor];
    self.shapeLayer.path = [[UIBezierPath bezierPathWithRoundedRect:self.bounds cornerRadius:5] CGPath];
}

+ (Class)layerClass {
    
    return [CAShapeLayer class];
}

- (CAShapeLayer *)shapeLayer {
    
    return (CAShapeLayer *)self.layer;
}

- (void)showLoading {
    
    if (self.activity) return;
//    
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"path"];
    
    animation.fromValue = (__bridge id _Nullable)
    ([[UIBezierPath bezierPathWithRoundedRect:self.bounds cornerRadius:5] CGPath]);
    
    animation.repeatCount = 1;
    animation.duration = 0.15;
    
    CGFloat r = MIN(self.frame.size.height, self.frame.size.height);
    animation.toValue = (__bridge id)
    [[UIBezierPath bezierPathWithRoundedRect:CGRectMake(self.frame.size.width/2 - r/2, 0, r, r)
                                cornerRadius:r] CGPath];
    
    [self.shapeLayer addAnimation:animation
                           forKey:@"shapeAnimation"];
    
    self.shapeLayer.path =
    [[UIBezierPath bezierPathWithRoundedRect:CGRectMake(self.frame.size.width/2 - r/2, 0, r, r)
                                cornerRadius:r] CGPath];
    
    [self showAtivityIndicator];
    self.currentText = self.currentTitle;
    [self setTitle:@"" forState:UIControlStateNormal];

    UIColor *fromColor = [UIColor colorWithRed:0.0392 green:0.3765 blue:1.0 alpha:1.0];
    UIColor *toColor =  [UIColor colorWithRed:0.0802 green:0.616 blue:0.1214 alpha:1.0];
    
    CABasicAnimation *colorAnimation = [CABasicAnimation animationWithKeyPath:@"fillColor"];
    colorAnimation.fromValue = (id)fromColor.CGColor;
    colorAnimation.toValue = (id)toColor.CGColor;
    colorAnimation.repeatCount = NSIntegerMax;
    colorAnimation.duration = 1.0;
    colorAnimation.autoreverses = YES;
    
    [self.shapeLayer addAnimation:colorAnimation forKey:@"color"];
}

- (void)hideLoading {
    
    if (!self.activity) return;
    
    self.shapeLayer.path = [[UIBezierPath bezierPathWithRoundedRect:self.bounds cornerRadius:5] CGPath];
    self.shapeLayer.fillColor = [[UIColor colorWithRed:0.0392 green:0.3765 blue:1.0 alpha:1.0] CGColor];
    
    [self hideActivityIndicator];
    [self setTitle:self.currentText forState:UIControlStateNormal];
    self.currentText = nil;
}

- (void)showAtivityIndicator {
    
    if (!self.activity) {
        
        self.userInteractionEnabled = NO;
        self.activity = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:(UIActivityIndicatorViewStyleWhite)];
        [self.activity setHidden:NO];
        [self.activity startAnimating];
        [self.activity setCenter:CGPointMake(self.frame.size.width/2, self.frame.size.height/2)];
        [self addSubview:self.activity];
    }
}

- (void)hideActivityIndicator {
    
    self.userInteractionEnabled = YES;
    [self.shapeLayer removeAllAnimations];
    [self.activity removeFromSuperview];
    self.activity = nil;
}

- (void)setEnabled:(BOOL)enabled {
    
    [super setEnabled:enabled];
    
    if (enabled) {
        
        self.shapeLayer.fillColor = [[UIColor colorWithRed:0.0392 green:0.3765 blue:1.0 alpha:1.0] CGColor];
    }
    else {
        
        self.shapeLayer.fillColor = [[UIColor grayColor] CGColor];
    }
}

@end

File: sample-conference-videochat-swift/sample-conference-videochat-swift/sample-conference-videochat-swift/Components/Auth/Views/LoadingButton/QBLoadingButton.swift
//
//  QBLoadingButton.swift
//  sample-conference-videochat-swift
//
//  Created by Vladimir Nybozhinsky on 04.10.2018.
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit

class QBLoadingButton: UIButton {
    
    var path: UIBezierPath!

    lazy private var shapeLayer: CAShapeLayer = {
        let shapeLayer = CAShapeLayer()
        return shapeLayer
    }()
    lazy private var activity: UIActivityIndicatorView = {
        let activity = UIActivityIndicatorView(style: .white)
        activity.hidesWhenStopped = true
        return activity
    }()
    private var currentText = ""
    
    class func layerClass() -> AnyClass {
        return CAShapeLayer.self
    }
    override var isEnabled: Bool{
        didSet {
            updateEnabled(isEnabled)
        }
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        isEnabled = false
        setTitle(currentText, for: .normal)
        shapeLayer.path = UIBezierPath(roundedRect: bounds, cornerRadius: 5).cgPath
        layer.addSublayer(shapeLayer)
    }

    public func showLoading() {
        if activity.isAnimating {
            return
        }
        let animation = CABasicAnimation(keyPath: "path")
        animation.fromValue = UIBezierPath(roundedRect: bounds, cornerRadius: 5).cgPath
        animation.repeatCount = 1
        animation.duration = 0.15
        let r = min(frame.size.height, frame.size.height)
        animation.toValue = (UIBezierPath(roundedRect: CGRect(x: frame.size.width / 2 - r / 2,
                                                              y: 0, width: r, height: r),
                                          cornerRadius: r)).cgPath
        shapeLayer.add(animation, forKey: "shapeAnimation")
        shapeLayer.path = (UIBezierPath(roundedRect: CGRect(x: frame.size.width / 2 - r / 2,
                                                            y: 0, width: r, height: r),
                                        cornerRadius: r)).cgPath
        
        showAtivityIndicator()
        currentText = self.currentTitle ?? "Title"
        self.setTitle("", for: .normal)
        
        let fromColor = UIColor(red: 0.0392, green: 0.3765, blue: 1.0, alpha: 1.0)
        let toColor = UIColor(red: 0.0802, green: 0.616, blue: 0.1214, alpha: 1.0)
        
        let colorAnimation = CABasicAnimation(keyPath: "fillColor")
        colorAnimation.fromValue = fromColor.cgColor
        colorAnimation.toValue = toColor.cgColor
        colorAnimation.repeatCount = Float(NSIntegerMax)
        colorAnimation.duration = 1.0
        colorAnimation.autoreverses = true
        
        shapeLayer.add(colorAnimation, forKey: "color")
    }
    
    public func hideLoading() {
        if activity.isAnimating == false {
            return
        }
        shapeLayer.path = UIBezierPath(roundedRect: bounds, cornerRadius: 5).cgPath
        shapeLayer.fillColor = UIColor(red: 0.0392, green: 0.3765, blue: 1.0, alpha: 1.0).cgColor
        
        hideActivityIndicator()
        self.setTitle(currentText, for: .normal)
        currentText = ""
    }
    
    private func showAtivityIndicator() {
        if activity.isAnimating == false {
            isUserInteractionEnabled = false
            activity.isHidden = false
            activity.startAnimating()
            activity.center = CGPoint(x: frame.size.width / 2, y: frame.size.height / 2)
            addSubview(activity)
        }
    }
    
    private func hideActivityIndicator() {
        isUserInteractionEnabled = true
        shapeLayer.removeAllAnimations()
        activity.removeFromSuperview()
    }
    
    private func updateEnabled(_ enabled: Bool) {
        if enabled {
            shapeLayer.fillColor = UIColor(red: 0.0392, green: 0.3765, blue: 1.0, alpha: 1.0).cgColor
        } else {
            shapeLayer.fillColor = UIColor.gray.cgColor
        }
    }
}
